src/Serialize.h

```diff
@@ -33,6 +33,98 @@ namespace drafter {
 
     refract::Registry& GetNamedTypesRegistry();
 
+    template<typename T>
+    struct SectionInfo {
+        typedef T SectionType;
+        typedef snowcrash::SourceMap<SectionType> SourceMapType;
+
+        const SectionType& section;
+        const SourceMapType& sourceMap;
+        const bool empty;
+
+        SectionInfo(const SectionType& section, const SourceMapType& sourceMap) : section(section), sourceMap(sourceMap), empty(false) {}
+        SectionInfo() : section(SectionType()), sourceMap(SourceMapType()), empty(true) {}
+
+        /**
+         * BE CAREFUL while assign SectionInfo it probably will not work as you expected
+         * but we need this to allow store SectionInfo in containers
+         *
+         * alternative solution is store `section` and `sourceMap` in C++11 smart pointers
+         */
+        SectionInfo<T>& operator=(const SectionInfo<T>& other) { 
+            return *this; 
+        }
+
+        static const SourceMapType& NullSourceMap() {
+            static const SourceMapType nullSourceMap;
+            return nullSourceMap;
+        }
+
+        bool isNull() const { 
+            return empty; 
+        }
+
+        bool hasSourceMap() const {
+            const SourceMapType& null = NullSourceMap();
+            return &sourceMap != &null;
+        }
+    };
+
+    template <typename T>
+    SectionInfo<T> MakeSectionInfo(const T& section, const snowcrash::SourceMap<T>& sourceMap, const bool hasSourceMap)
+    {
+        return SectionInfo<T>(section, hasSourceMap ? sourceMap : SectionInfo<T>::NullSourceMap());
+    }
+
+
+    template <typename T>
+    SectionInfo<T> MakeSectionInfoWithoutSourceMap(const T& section)
+    {
+        return SectionInfo<T>(section, SectionInfo<T>::NullSourceMap());
+    }
+
+#define MAKE_SECTION_INFO(from, member) MakeSectionInfo(from.section.member, from.sourceMap.member, from.hasSourceMap())
```

Comment:
Couldn't we have moved this to the top of the file since this is a macro? We don't need this to be here.